tls/s2n_ktls.c

```diff
@@ -297,3 +308,60 @@ int s2n_config_ktls_enable_unsafe_tls13(struct s2n_config *config)
     config->ktls_tls13_enabled = true;
     return S2N_SUCCESS;
 }
+
+S2N_RESULT s2n_ktls_key_update_send(struct s2n_connection *conn)
+{
+    RESULT_ENSURE_REF(conn);
+
+    if (s2n_atomic_flag_test(&conn->key_update_pending)) {
+        uint8_t key_update_data[S2N_KEY_UPDATE_MESSAGE_SIZE];
+        struct s2n_blob key_update_blob = { 0 };
+        RESULT_GUARD_POSIX(s2n_blob_init(&key_update_blob, key_update_data, sizeof(key_update_data)));
+
+        /* Write Keyupdate message */
+        RESULT_GUARD_POSIX(s2n_key_update_write(&key_update_blob));
+
+        /* Send Keyupdate message */
+        const struct iovec iov = {
+            .iov_base = (void *) (uintptr_t) key_update_data,
+            .iov_len = sizeof(key_update_data),
+        };
+        s2n_blocked_status blocked = S2N_NOT_BLOCKED;
+        size_t bytes_written = 0;
+        RESULT_GUARD(s2n_ktls_sendmsg(conn->send_io_context, TLS_HANDSHAKE, &iov, 1, &blocked, &bytes_written));
+        RESULT_ENSURE_EQ(bytes_written, sizeof(key_update_data));
+
+        /* Create new encryption key */
+        RESULT_GUARD_POSIX(s2n_update_application_traffic_keys(conn, conn->mode, SENDING));
+
+        struct s2n_ktls_crypto_info crypto_info = { 0 };
+        RESULT_GUARD(s2n_ktls_crypto_info_init(conn, S2N_KTLS_MODE_SEND, &crypto_info));
+
+        int fd = 0;
+        RESULT_GUARD(s2n_ktls_get_file_descriptor(conn, S2N_KTLS_MODE_SEND, &fd));
+
+        /* Update the socket with new key */
+        int ret = s2n_setsockopt(fd, S2N_SOL_TLS, S2N_TLS_TX, crypto_info.value.data, crypto_info.value.size);
+        RESULT_ENSURE(ret == 0, S2N_ERR_KTLS_SOCKOPT);
+
+        s2n_atomic_flag_clear(&conn->key_update_pending);
+    }
+    return S2N_RESULT_OK;
+}
+
+S2N_RESULT s2n_ktls_key_update_recv(struct s2n_connection *conn)
+{
```

Comment:
Nit: naming wise, I think it's a little wonky that key_update_send sends a message and key_update_recv doesn't recv a message. Maybe process instead of recv?